connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -740,15 +755,26 @@ private void backoff(long ms) {
         Utils.sleep(ms);
     }
 
+    private void bulkRun(Collection<Callable<Void>> bulk) {
+        try {
+            bulkExecutor.invokeAll(bulk);
+        } catch (InterruptedException e) {
+            // ignore
+        }
+    }
+
     private void startWork() {
         // Start assigned connectors and tasks
         log.info("Starting connectors and tasks using config offset {}", assignment.offset());
+        List<Callable<Void>> callables = new ArrayList<>();
         for (String connectorName : assignment.connectors()) {
-            startConnector(connectorName);
+            callables.add(getConnectorCallable(connectorName, ConnectorStatus.State.RUNNING));
         }
+
         for (ConnectorTaskId taskId : assignment.tasks()) {
-            startTask(taskId);
+            callables.add(getTaskCallable(taskId, TaskStatus.State.RUNNING));
         }
+        bulkRun(callables);
```

Comment:
There has been some provision to make some of the above classes/methods thread safe. But this effort doesn't look complete overall. Plus, we need to double check that synchronization is composable across the workflow of the worker/herder. 

My rough suggestion would be to leave `KafkaBasedLog` with no thread-safety guarantees (in the same way the consumer is not thread safe) but make the implementations of the `StatusBackingStore` interface thread safe (and warn against non-thread safe implementations). Additionally make the whole workflow of connector reconfiguration thread-safe, mainly under the the presence of callbacks and asynchronous execution.